### PR TITLE
🐛 Add wait for MachineList to be available

### PR DESCRIPTION
**Pull request by adilGhaffarDev**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
trying to fix the flake mentioned here: https://github.com/kubernetes-sigs/cluster-api/issues/9688#issuecomment-1988364796

### CODE REVIEW

File: test/e2e/clusterctl_upgrade.go [hunk 1]
```diff
@@ -548,6 +548,24 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 				input.PostUpgrade(managementClusterProxy, workloadClusterNamespace, managementClusterName)
 			}
 
+			// After the upgrade check that MachineList is available.
```
Comment (killianmuldoon) [inline test/e2e/clusterctl_upgrade.go:551]: Could we add a comment about what problem we're solving here.
```suggestion
			// After the upgrade check that MachineList is available. This ensures the APIServer is serving without error before checking that it `Consistently` returns the MachineList later on.
```

**Review by killianmuldoon — COMMENTED** [review]
/lgtm

Thanks!

**Comment by adilGhaffarDev** [comment]
> Test name 	Commit 	Details 	Required 	Rerun command
> pull-cluster-api-e2e-main 	[882933e](https://github.com/kubernetes-sigs/cluster-api/commit/882933e363ff779a7d3684ead57fa0d83715d241) 	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/10301/pull-cluster-api-e2e-main/1772274269649113088) 	true 	`/test pull-cluster-api-e2e-main`

failed because of other flake.

**Comment by adilGhaffarDev** [comment]
@sbueringer can we merge this before today's release if it looks ok?

**Comment by chrischdi** [comment]
/lgtm

Thx for adding this. This makes sense!

/assign sbueringer

**Comment by sbueringer** [comment]
Yup let's merge it

/approve

**Pull request merged**